flocker/node/agents/blockdevice.py

```diff
@@ -1468,6 +1540,94 @@ def _get_system_mounts(self, volumes, compute_instance_id):
+            FilePath(partition.device)
+            for partition
+            in psutil.disk_partitions()
+        }
+
+        compute_instance_id = api.compute_instance_id()
+        for volume in _attached_volumes(compute_instance_id, volumes):
+            try:
+                get_device_for_dataset_id(volume.dataset_id)
+                # If we found the dataset_id like this the filesystem is
+                # sufficiently new to not need any changes.  Proceed to the
+                # next volume.
+                continue
+            except KeyError:
+                # There is no filesystem on the volume because volume setup
+                # failed partway through the process or there is no filesystem
+                # with a UUID that matches the dataset id because the volume
+                # was created by 1.0.0.
+                #
+                # The filesystem isn't mounted because the 1.0.1 upgrade
+                # instructions tell the user to restart the node upon upgrade.
+                #
+                # Detach and then re-attach the volume so that get_device_path
+                # can tell us the device.  This is slow and we might be able to
+                # save time by calling get_device_path first.  It might give us
+                # an answer.  Maybe do that optimization.
+                api.detach_volume(volume.blockdevice_id)
+                api.attach_volume(volume.blockdevice_id, compute_instance_id)
+                device_path = api.get_device_path(volume.blockdevice_id)
+
+            # Find out of it is mounted.  Skip it (don't break the rest of the
+            # loop) if it is mounted because there's nothing we can do about
+            # mounted filesystems.  They're not supposed to be mounted because
+            # we told the user to restart the node.
+            if device_path in mounted:
+                FLOCKER_1_0_0_FS_MOUNTED(
+                    block_device_id=volume.blockdevice_id,
+                    block_device_path=device_path,
+                ).write()
+                continue
+
+            try:
+                old_fs_uuid = _ext4_fs_uuid(device_path)
+            except FilesystemMissing:
+                # The volume was most likely only partially initialized and
+                # then the system failed in some way leaving it without the
+                # filesystem it's meant to have.  This would be a good time
+                # to mark the volume for fixing (put a filesystem on it).
+                # For now, ignore it.
+                #
+                # FLOC-1576
+                continue
+
+            # Put the dataset id into the ext4 filesystem UUID field for future
+            # reference.
+            with FLOCKER_1_0_0_FS_UPGRADE(
+                block_device_id=volume.blockdevice_id,
+                dataset_id=volume.dataset_id
+            ) as upgrade:
+                upgrade.addSuccessFields(old_fs_uuid=unicode(old_fs_uuid))
```

Comment:
:question: I wonder if the `clusterhq-flocker-node` package has an `e2fsprogs` dependency?